modules/openthread/platform/spinel_hdlc.cpp

```diff
@@ -0,0 +1,411 @@
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#include <assert.h>
+#include <openthread/tasklet.h>
+#include <openthread/platform/alarm-milli.h>
+#include <common/logging.hpp>
+#include <common/code_utils.hpp>
+#include "spinel_hdlc.hpp"
+
+namespace ot {
+
+namespace Zephyr {
+
+HdlcInterface::HdlcInterface(const Url::Url &aRadioUrl)
+    : mEncoderBuffer()
+    , mHdlcEncoder(mEncoderBuffer)
+    , mHdlcRxCallbackField(nullptr)
+    , mReceiveFrameBuffer(nullptr)
+    , mReceiveFrameCallback(nullptr)
+    , mReceiveFrameContext(nullptr)
+    , mHdlcSpinelDecoder()
+    , mIsInitialized(false)
+    , mSavedFrame(nullptr)
+    , mSavedFrameLen(0)
+    , mIsSpinelBufferFull(false)
+    , mRadioUrl(aRadioUrl)
+{
+    mHdlcRxCallbackField = HdlcRxCallback;
+
+    radio_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_hdlc_rcp_if));
+    __ASSERT_NO_MSG(device_is_ready(radio_dev));
+
+	hdlc_api = (struct hdlc_api *)radio_dev->api;
+	if (!hdlc_api)
+    {
+        otLogDebgPlat("Radio device initialization failed");
+        assert(0);
+    }
+}
+
+HdlcInterface::~HdlcInterface(void)
+{
+}
+
+otError HdlcInterface::Init(ReceiveFrameCallback aCallback, void *aCallbackContext, RxFrameBuffer &aFrameBuffer)
+{
+    int status;
+    otError error = OT_ERROR_NONE;
+
+    if (!mIsInitialized)
+    {
+        /* Event initialization */
+        k_event_init(&s_spinel_hdlc_event);
+
+        /* Read/Write semaphores initialization */
+        k_mutex_init(&s_spinel_hdlc_wr_lock);
+        k_mutex_init(&s_spinel_hdlc_rd_lock);
+
+        spinel_hdlc_msgq_buffer = (char *)k_malloc(1 * sizeof(spinel_hdlc_msgq_buffer));
```

Comment:
Please avoid using dynamic allocaton if not necessary.

Reply:
resolved